### 📚 구체적인 과제 수행 과정

#### 1️⃣ 데이터셋 탐색

- pandas 라이브러리의 read 함수 사용하여 데이터 파일을 데이터프레임으로 불러오기

In [63]:
# 데이터 불러오기
import pandas as pd

# csv 파일 불러오기 (./는 현재 디렉토리에 있는 파일을 가져오겠다는 뜻이다.)
df = pd.read_csv('./supervised_learning_data/housingdata.csv')

# 전체 데이터프레임 확인
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93   0.0  0.573  6.030   NaN  2.5050    1  273   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90    NaN  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99    NaN  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [64]:
print('housingdata.csv 데이터 탐색')
print('-'*130)

# 데이터프레임의 정보 확인 (컬럼, 결측값, 데이터 타입 등)
df.info()
print('-'*130)

print(f'컬럼 : {list(df.columns)}')
print('-'*130)

print(f'데이터프레임의 크기 (행,열) : {df.shape}')
print('-'*130)

# 데이터프레임의 요약 통계량 확인
df.describe() 

housingdata.csv 데이터 탐색
----------------------------------------------------------------------------------------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     486 non-null    float64
 1   ZN       486 non-null    float64
 2   INDUS    486 non-null    float64
 3   CHAS     486 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      486 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    486 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB
---------------------------------------------------------------------------------

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  486.000000  486.000000  486.000000  486.000000  506.000000  506.000000   
mean     3.611874   11.211934   11.083992    0.069959    0.554695    6.284634   
std      8.720192   23.388876    6.835896    0.255340    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.081900    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.253715    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.560263   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  486.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.518519    3.795043    9.549407  408.237154   18.455534  356.674032   
std     27.999513    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.175000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     76.800000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     93.975000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  486.000000  506.000000  
mean    12.715432   22.532806  
std      7.155871    9.197104  
min      1.730000    5.000000  
25%      7.125000   17.025000  
50%     11.430000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

##### 💡 housingdata.csv 파일을 데이터프레임으로 만든 후 여러 메서드를 사용해 탐색한 결과

1. 506행, 14열의 데이터프레임이다.
2. 총 14개의 컬럼이 존재한다. ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV' ]
3. range index(연속 정수형 인덱스)가 총 506개, 즉, 0부터 505인덱스까지 있음을 알 수 있다.
4. 따라서, Non-Null Count가 506 non-null이 아닌 ['CRIM', 'ZN', 'INDUS', 'CHAS','AGE','LSTAT'] 컬럼은 결측값이 있는 것임을 알 수 있다. 
5. 'RAD', 'TAX'는 정수형, 그 외는 모두 부동 소수점 데이터 타입이다.

---

#### 2️⃣ 데이터셋 전처리 (결측치 처리, 이상치 탐지 및 제거, 특징 선택)
##### 💡 전처리는 데이터 분석 및 머신러닝 모델링을 위해 데이터를 준비하는 과정으로, 데이터의 품질을 높이고 분석 결과의 신뢰성을 확보하기 위한 필수적인 과정 
##### 📝 결측치 처리 (데이터셋에서 누락된 값을 처리하는 작업) : 결측치를 처리하지 않으면 모델의 성능이 저하될 수 있다.

- 결측값이나 이상치는 데이터의 품질을 저하시킬 수 있다. 데이터 품질이 낮으면 모델의 성능이 떨어질 수 있고, 잘못된 예측을 할 가능성이 높아진다.
- 결측값이 포함된 데이터는 계산 과정에서 오류를 발생시킬 수 있으며, 이상치는 모델의 파라미터에 큰 영향을 미쳐 잘못된 학습 결과를 초래할 수 있다.
- 결측값과 이상치를 적절히 처리하면 모델의 일반화 능력과 예측 성능이 향상될 수 있다. 이는 모델이 새로운 데이터에 대해 더 나은 예측을 할 수 있게 해준다.

    - 제거 : 결측치가 있는 행이나 열을 제거한다. 정확한 데이터만 가질 수 있어 결측치가 적을 때는 유용하지만, **데이터 손실**이 발생할 수 있다.
    - 대체 : 평균, 중앙값, 최빈값 등으로 결측치를 대체한다.
    - 예측 : 다른 특성을 사용하여 결측치를 예측하고 채운다.  

➡️ 해당 데이터의 특성에 맞게 결측치를 처리해야 한다.**다양한 방법을 시도**해보고 가장 적합한 방법을 선택하는 것이 중요하다.

##### 📝 현재 결측치가 있는 컬럼들

- **CRIM**: 마을별 1인당 범죄율 [mean : 3.611874 / min : 0.006320 / max : 88.976200]
    - 최댓값이 크긴하지만 평균값을 봤을 때 대체로 작은 값에 몰려있는 것 같아, 평균값으로 대체해주겠다.
- **ZN**: 25,000평방피트 이상의 부지에 대해 구획된 주거용 토지의 비율 [mean : 11.211934 / min : 0.000000 / max : 100.000000]
    - 데이터를 봤을 때 0이 이어지다가 어디는 수치가 몰려있는, 군집화된 양상을 띄고 있다. 비어있는 결측값에 아무거나 넣어놓으면 군집화가 깨질 수도 있을 것 같아, 결측값이 있는 행은 제거하겠다.
- **INDUS**: 마을별 비소매업무 지역 비율 [mean : 11.083992 / min : 0.460000 / max : 27.740000]
    - 값이 최솟값부터 최댓값까지 고루 분포하는 양상을 보여서, 평균값으로 대체해주겠다.
- **CHAS**: 찰스 강 더미 변수 (강이 인접한 경우 1, 그렇지 않으면 0) [mean : 0.069959 / min : 0.000000 / max : 1.000000]
    - 더미형 데이터의 경우 결측값은 날려버리는 것도 괜찮다고 해서 결측값이 있는 행을 제거하겠다. 
- **AGE**: 1940년 이전에 지어진 소유주 점유 비율 [mean : 68.518519 / min : 2.900000 / max : 100.000000]
    - 값이 고루 분포하는 양상을 보여서, 평균값으로 대체해주겠다. 
- **LSTAT**: 인구 중 저소득층 비율 [mean : 12.715432 / min : 1.730000 / max : 37.970000]
    - 값이 고루 분포하는 양상을 보여서, 평균값으로 대체해주겠다.

➡️ 선형회귀 모델 등을 사용하여 결측값을 예측한 후 넣어줄 수도 있으나 일단 대체하는 방법으로 과제를 진행해보겠다.

In [65]:
# 데이터셋의 결측값을 다른 값으로 대체하기
# inplace = True 옵션으로 원본데이터 변경

df['CRIM'].fillna(df['CRIM'].mean(), inplace=True) # 평균값
df['INDUS'].fillna(df['INDUS'].mean(), inplace=True) # 평균값
df['AGE'].fillna(df['AGE'].mean(), inplace=True) # 평균값
df['LSTAT'].fillna(df['LSTAT'].mean(), inplace=True) # 평균값

# 그 외 결측값이 있는 행은 제거하기
df.dropna(inplace=True)

print('각 컬럼들의 결측값 현황')
print(df.isna().sum())  # 결측값이 잘 처리되었는지 확인 (결측값을 대체하거나 결측값이 있는 행을 제거해서 각 컬럼의 결측값이 0인 것을 알 수 있다.)
print('-'*130)
df.info() # 데이터프레임 정보 다시 확인 (총 506행의 데이터에서 466행의 결측값 없는 데이터로 바뀐 것을 확인 가능) (❗️ 결측값이 있는 행을 제거했을 뿐 인덱스는 기존과 같이 남아있다는 것 유의)

각 컬럼들의 결측값 현황
CRIM       0
ZN         0
INDUS      0
CHAS       0
NOX        0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64
----------------------------------------------------------------------------------------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
Index: 466 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     466 non-null    float64
 1   ZN       466 non-null    float64
 2   INDUS    466 non-null    float64
 3   CHAS     466 non-null    float64
 4   NOX      466 non-null    float64
 5   RM       466 non-null    float64
 6   AGE      466 non-null    float64
 7   DIS      466 non-null    float64
 8   RAD      466 non-null    int64  
 9   TAX      466 non-null    int64  
 10  PTRATIO  466 non-null    float64
 11  B        466 non-null    float64
 12  LSTAT    466 non-null  

##### 📝 이상치(데이터에서 비정상적으로 크거나 작은 값) 탐지 및 제거 : 이상치는 분석결과에 큰 영향을 미쳐 모델의 성능을 저하시킬 수 있다.
- 제거 : 이상치를 데이터셋에서 제거한다.
- 대체 : 이상치를 특정 값으로 대체한다. 
- 변환 : 이상치를 변환하여 데이터의 분포를 조정한다. (예: 상한선이나 하한선으로 대체)
- IQR 방법 : IQR(interquartile Range : 사분위 범위)을 사용하여 이상치를 탐지하고 처리한다. (**정수형 및 실수형 데이터**)
    - IQR 방법은 데이터의 범위와 분포를 이해하고 이상값(outliers)을 탐지하는 데 사용되는 통계적 기법이다. IQR은 데이터셋에서 상위 25%를 제거한 중간 50%의 범위를 나타낸다. 이는 데이터의 중간값 주변의 분포를 분석할 때 유용하다.
    - 백분위수 : 데이터를 백분위수로 나누었을 때의 값을 나타내는 통계적 척도. 백분위수는 데이터를 100개의 동일한 부분으로 나누는 방식으로, 특정 백분위수는 전테 데이터의 몇 퍼센트가 해당 값 이하인지를 나타낸다. 
    - 백분위수는 데이터의 분포를 이해하고 데이터 간의 상대적 위치를 파악하는 데 유용하게 사용된다.
        - 0 백분위수 : 최솟값
        - 25 백분위수 : 1사분위수(Q1), 전체 데이터의 하위 25%가 이 값 이하
        - 50 백분위수 : 중앙값(median), 전체 데이터의 하위 50%가 이 값 이하
        - 75 백분위수 : 3사분위수(Q3), 전체 데이터의 하위 75%가 이 값 이하
        - 100 백분위수 : 최댓값
        - IQR : Q3 - Q1, 즉 3사분위수에서 1사분위수를 뺀 값이다.
    - IQR 방법을 사용한 이상값 탐지 : 이상값은 일반적으로 다음 범위를 벗어난 값으로 정의 된다.
        - 하한값(lower_bound) : Q1 - 1.5 * IQR
        - 상한값(upper_bound) : Q3 + 1.5 * IQR

In [66]:
#  IQR 방법을 활용한 각 컬럼의 이상치 확인

# 특정 열의 이상치 확인 (IQR 방법)
# 데이터프레임의 quantile 메서드는 특정 백분위수(quantile)에 해당하는 값을 셰산하는 데 사용된다. 주어진 백분위수를 기준으로 데이터를 나눠주는 값을 반환한다.
Q1 = df['CRIM'].quantile(0.25)
Q3 = df['CRIM'].quantile(0.75)
IQR = Q3 - Q1

# 이상치 범위 설정
lower_bound = Q1 - 1.5 * IQR # 하한값
upper_bound = Q3 + 1.5 * IQR # 상한값

In [67]:
# 해당 컬럼에서 이상치가 있는 행 확인해보기 (하한값보다 작거나 상한값보다 큰 값은 이상치로 간주한다.)
outliers = df[(df['CRIM'] < lower_bound) | (df['CRIM'] > upper_bound)]
print(outliers) # 이상치로 간주되는 데이터 포인트들을 보여준다.

         CRIM   ZN      INDUS  CHAS    NOX     RM         AGE     DIS  RAD  \
356   8.98296  0.0  18.100000   1.0  0.770  6.212   97.400000  2.1222   24   
371   9.23230  0.0  18.100000   0.0  0.631  6.216  100.000000  1.1691   24   
373  11.10810  0.0  18.100000   0.0  0.668  4.906  100.000000  1.1742   24   
374  18.49820  0.0  18.100000   0.0  0.668  4.138  100.000000  1.1370   24   
377   9.82349  0.0  18.100000   0.0  0.671  6.794   98.800000  1.3580   24   
378  23.64820  0.0  18.100000   0.0  0.671  6.380   96.200000  1.3861   24   
379  17.86670  0.0  18.100000   0.0  0.671  6.223  100.000000  1.3861   24   
380  88.97620  0.0  18.100000   0.0  0.671  6.968   91.900000  1.4165   24   
381  15.87440  0.0  18.100000   0.0  0.671  6.545   99.100000  1.5192   24   
382   9.18702  0.0  18.100000   0.0  0.700  5.536  100.000000  1.5804   24   
384  20.08490  0.0  18.100000   0.0  0.700  4.368   91.200000  1.4395   24   
385  16.81180  0.0  18.100000   0.0  0.700  5.277   98.100000  1

In [68]:
# 이상치 처리
# 이상치를 평균값으로 대체
mean_value = df['CRIM'].mean()
# 해당 열 전체의 각 데이터 포인터를 검사하여 이상치인 경우 평균값으로 대체하고, 그렇지 않은 경우 원래 값을 유지한다.
df['CRIM'] = df['CRIM'].apply(lambda x: mean_value if x < lower_bound or x > upper_bound else x)

outliers = df[(df['CRIM'] < lower_bound) | (df['CRIM'] > upper_bound)]
print(outliers) # 이상치를 처리가 됐는지 확인

Empty DataFrame
Columns: [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD, TAX, PTRATIO, B, LSTAT, MEDV]
Index: []


이제 이 과정을 나머지 모든 컬럼에 반복해주면 데이터셋 전체의 이상치 처리 완료.

In [69]:
# 해당 과정을 나머지 모든 컬럼에 반복

columns = ['ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']

for column in columns:
    Q1 = df[column].quantile(0.25) # 1사분위
    Q3 = df[column].quantile(0.75) # 3사분위
    IQR = Q3 - Q1 # IQR
    
    # 이상치 탐지 범위 설정
    lower_bound = Q1 - 1.5 * IQR # 하한값
    upper_bound = Q3 + 1.5 * IQR # 상한값

    # 컬럼의 평균값 계산
    mean_value = df[column].mean()
    
    # 이상치를 평균값으로 대체
    df[column] = df[column].apply(lambda x: mean_value if x < lower_bound or x > upper_bound else x)

이제 결측값과 이상치를 처리했으니 머신러닝 모델을 위한 특징을 선택하자!

#### 📝 특징 선택(Feature Selection) 및 추출(Feature Extraction)

- 머신러닝 모델 성능을 높이기 위해 중요한 특징을 선택하거나 새로운 특징을 추출하는 과정. 
- 특징 선택 과정을 통해 모델 학습에 가장 유용한 특징들을 선별하고, 이를 포함하는 새로운 입력 데이터를 만들어 모델을 학습시키는 것이다. 이런 접근법은 모델의 성능을 최적화하고 계산 비용을 줄여줄 수 있다.
- 한 마디로, 특징을 선택한다는 것은 기존 입력 데이터에서, 특정 컬럼을 선택하여 모델 학습에 가장 유용한 특징(컬럼)만을 포함하는 새로운 입력 데이터를 만든다는 뜻이다. 불필요한 특징을 제거하여 계산 비용을 줄임!

다음은 scikit-learn 라이브러리의 SelectKBest 클래스를 사용하여 상위 5개의 중요한 특징을 선택하는 과정이다.

#### **SelectKBest 클래스**

- scikit-learn 라이브러리의 SelectKBest 클래스는 가장 좋은 K개의 특징을 선택하는 데 사용된다.
- 여기서 '가장 좋은'을 결정하는 기준은 특징 선택 방법에 따라 달라지는데, 여러 가지 방법이 있지만 SelectKBest와 f_classif를 사용하는 경우 '가장 좋은' 특징을 결정하는 기준은 ANOVA F-검정(Analysis of Variance F-test)을 기반으로 한다.

#### **ANOVA F-검정**

- 목적: **각 특징이 타겟 변수와 얼마나 강한 상관관계를 가지는지** 평가한다.
- F-통계량: 각 특징의 설명력이 타겟 변수에 얼마나 기여하는지를 나타내는 통계량이다.
    - 높은 F-통계량: 해당 특징이 타겟 변수와 강한 상관관계를 가지며, '좋은' 특징으로 간주된다.

- 특징 선택 과정
    - 특징 평가: 각 특징에 대해 ANOVA F-검정을 수행하여 F-통계량을 계산한다.
    - 순위 매기기: F-통계량에 따라 특징을 내림차순으로 정렬한다.
    - 특징 선택: 가장 높은 F-통계량을 가진 상위 K개의 특징을 선택한다. 
  
#### **f_classif**

- 분산분석(ANOVA) F-검정을 수행하는 함수로, **각 특징의 중요도를 평가**한다.

In [70]:
# 모델에서 중요한 특징을 선택하는 과정

# 특징 선택방법 설정
from sklearn.feature_selection import SelectKBest, f_classif

# 입력 데이터(X)와 타겟 변수(y) 설정 
# 예시로 'MEDV'(주택 가격 데이터셋에서 주택 가격의 중간값)를 타겟 변수로 설정. 즉, 'MEDV'는 예측하고자 하는 타겟 변수이기 때문에 입력데이터에서 제외해야 한다.(정답지 주고 학습시키면 안됨!)

# X : 입력 데이터. 모델이 학습하는데 사용하는 특징(features)들을 담고 있는 데이터 프레임
# 타겟변수인 'MEDV'를 제외한 모든 열이 포함된 데이터프레임
X = df.drop(columns=['MEDV']) 

# y : 예측해야 하는 목표값을 담고 있는 시리즈 또는 벡터
y = df['MEDV'] # 데이터프레임에서 MEDV 열을 선택하여 타겟변수로 설정한다.


# SelectKBest 객체 생성(각 특징의 중요도를 평가할 함수로 f_classif를 사용하고, 상위 5개의 특징을 선택한다.)
# 여기서 selector는 상위 5개의 특징을 선택할 수 있도록 하는 메커니즘을 설정하는 객체이다.
selector = SelectKBest(score_func = f_classif, k=5) 

# 데이터에 맞추기 (fit) 및 변환 (transform)

# fit_transform 메서드를 사용하여 입력 데이터 X와 타겟변수 y에 맞춰서 학습(fit)하고, 상위 5개의 중요한 특징을 선택하여 변환(transform)한다.
# 결과적으로 선택된 상위 5개의 특징이 포함된 새로운 데이터 X_new 가 생성된다.

# 입력 데이터 X와 타겟 변수 y를 사용하여 모델을 학습(fit)하고, 상위 5개의 특징이 포함된 새로운 데이터 X_new를 반환한다.
# fit_transform 메서드는 Scikit-learn과 같은 머신러닝 라이브러리에서 데이터 전처리와 특징 선택과 같은 작업을 수행한다.
# 먼저 데이터에 맞추어 모델이나 변환기를 학습(fit)한다. 이 과정에서는 모델이 데이터의 구조나 분포를 학습하게 된다.
# 학습된 모델이나 변환기를 사용하여 데이터를 변환(transform)합니다. 이 과정에서는 학습된 정보를 바탕으로 데이터를 실제로 변환하여 새로운 형식으로 반환합니다.
X_new = selector.fit_transform(X, y) 

# get_support 메서드는 sklearn의 특징 선택 클래스에서 사용되며, 선택된 특징의 인덱스를 반환한다.
# 선택된 특징의 인덱스를 반환한다. 이를 통해 어떤 특징(어떤 행)이 선택되었는지 확인할 수 있다.
# ndices=True 옵션은 메서드가, 선택된 특징의 인덱스를 반환하도록 지정한다. 선택된 특징의 인덱스는 해당 특징이 입력 데이터에서 몇 번째 '열'에 위치해 있는지를 나타낸다.
selected_features = selector.get_support(indices=True) 
print("선택된 특징의 인덱스:", selected_features) 
print("선택된 특징:", X.columns[selected_features]) # 선택된 상위 5개의 특징 컬럼을 보여줌

선택된 특징의 인덱스: [ 2  4  5  9 12]
선택된 특징: Index(['INDUS', 'NOX', 'RM', 'TAX', 'LSTAT'], dtype='object')


⭐️ X_new는 SelectKBest를 사용하여 기존의 입력 데이터 X에서 중요한 상위 K개의 특징만을 포함하는 새로운 입력 데이터인 것이다! ⭐️

X_new에는 선택된 상위 K개의 특징(컬럼)만 포함되어 있고, 나머지 특징들은 모두 제거된 상태다.

요약
- 기존의 입력 데이터 X : 원래의 모든 특징(컬럼)을 포함한 데이터프레임.
- 새로운 입력 데이터 X_new : SelectKBest에 의해 선택된 중요한 상위 K개의 특징(컬럼)만을 포함한 넘파이 배열(array). X_new는 X와는 달리 선택된 k개의 열만을 갖게 된다.

X_new는 fit_transform 메서드의 결과로 생성된 NumPy 배열(array)이다. 이는 모델 학습과 같은 머신러닝 작업에 최적화된 형식이다.

- NumPy 배열은 성능이 뛰어나고, 다차원 배열 연산을 쉽게 수행할 수 있기 때문에 머신러닝 모델 학습에 매우 적합하다.

In [71]:
# 새로운 입력 데이터 X_new 확인
X_new

array([[  2.31 ,   0.538,   6.575, 296.   ,   4.98 ],
       [  7.07 ,   0.469,   6.421, 242.   ,   9.14 ],
       [  7.07 ,   0.469,   7.185, 242.   ,   4.03 ],
       ...,
       [ 11.93 ,   0.573,   6.976, 273.   ,   5.64 ],
       [ 11.93 ,   0.573,   6.794, 273.   ,   6.48 ],
       [ 11.93 ,   0.573,   6.03 , 273.   ,   7.88 ]])

In [75]:
print(X_new.shape) # 여전히 466행이고, 상위 5개 특징 열만 뽑아서 5열인 배열임을 알 수 있다.

(466, 5)


In [74]:
# 배열을 데이터프레임으로 바꿔서 확인해보자
X_new_df = pd.DataFrame(X_new, columns=X.columns[selected_features]) 
print("X_new (데이터프레임):")
print(X_new_df)

X_new (데이터프레임):
     INDUS    NOX     RM    TAX      LSTAT
0     2.31  0.538  6.575  296.0   4.980000
1     7.07  0.469  6.421  242.0   9.140000
2     7.07  0.469  7.185  242.0   4.030000
3     2.18  0.458  6.998  222.0   2.940000
4     2.18  0.458  7.147  222.0  12.715432
..     ...    ...    ...    ...        ...
461  11.93  0.573  6.593  273.0  12.715432
462  11.93  0.573  6.120  273.0   9.080000
463  11.93  0.573  6.976  273.0   5.640000
464  11.93  0.573  6.794  273.0   6.480000
465  11.93  0.573  6.030  273.0   7.880000

[466 rows x 5 columns]


이제 특징 선택이 완료된 새로운 입력데이터 X_new까지 구해놨으니, 모델을 만들어보자 !